### PR TITLE
BUG: Fix metadata propagation in reductions

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -10940,12 +10940,13 @@ def any(  # type: ignore[override]
...
-    ) -> Series:
-        # error: Incompatible return value type (got "Union[Series, bool]",
-        # expected "Series")
-        return self._logical_func(  # type: ignore[return-value]
+    ) -> Series | bool:
+        result = self._logical_func(
             "any", nanops.nanany, axis, bool_only, skipna, **kwargs
         )
+        if isinstance(result, Series):
+            result = result.__finalize__(self, method="any")
```
[the remark this answers is not in the document]
Comment (lithomas1) [reply pandas/core/frame.py:10948]: Probably, but mypy won't be able to figure out that result is a Series without an isinstance/cast, though.

File: pandas/core/reshape/encoding.py [hunk 2]
```diff
@@ -4,6 +4,7 @@
 from collections.abc import (
     Hashable,
     Iterable,
+    cast,
 )
 import itertools
 from typing import TYPE_CHECKING
```
Comment (mroeschke) [inline pandas/core/reshape/encoding.py:10]:
```suggestion

)
import itertools
from typing import (
    TYPE_CHECKING,
    cast,
)
```

**Comment by mroeschke** [comment]
pre-commit.ci autofix

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @lithomas1

**Pull request merged**